src/librustc/infer/const_variable.rs

```diff
@@ -0,0 +1,271 @@
+use crate::mir::interpret::ConstValue;
+use syntax::symbol::InternedString;
+use syntax_pos::Span;
+use crate::ty::{self, InferConst};
+
+use std::cmp;
+use std::marker::PhantomData;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::snapshot_vec as sv;
+use rustc_data_structures::unify as ut;
+
+pub struct ConstVariableTable<'tcx> {
+    values: sv::SnapshotVec<Delegate<'tcx>>,
+
+    relations: ut::UnificationTable<ut::InPlace<ty::ConstVid<'tcx>>>,
+}
+
+/// Reasons to create a const inference variable
+#[derive(Copy, Clone, Debug)]
+pub enum ConstVariableOrigin {
+    MiscVariable(Span),
+    ConstInference(Span),
+    ConstParameterDefinition(Span, InternedString),
+    SubstitutionPlaceholder(Span),
```

Comment:
I just realized all of these have `Span`s, maybe that can be factored out? (but this is probably not the only place)

Reply:
I'll refactor `TypeVariableOrigin` and `ConstVariableOrigin` in a follow up.

Reply:
Opened https://github.com/rust-lang/rust/pull/59331.